### PR TITLE
Added USERNAME/PASSWORD to RedisServer init (jakub to dev)

### PR DESCRIPTION
**Pull request by ZaiZu2**

### CODE REVIEW

File: keepvariable/keepvariable_core.py [hunk 1]
```diff
@@ -196,9 +195,8 @@ def parse_saved_value(self, value, additional_params: Optional[dict] = None):
 
         return value
 
-    def decode_loaded_value(
-        self, value: str
-    ) -> Union[dict, pd.DataFrame, np.ndarray, datetime.datetime]:
+    def decode_loaded_value(self,
+                            value: str) -> Union[dict, pd.DataFrame, np.ndarray, datetime.datetime]:
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:227]: previous formating seemed nicer (ending bracket on a new line

File: keepvariable/keepvariable_core.py [hunk 2]
```diff
@@ -309,18 +319,20 @@ def release(self):
...
+        self.storage[name] = value
+        return {name: value}
+
+    # QUESTION: For in-app memory, is there a point to store serialized values?
+    def get(self, name: str) -> Union[dict, pd.DataFrame, np.ndarray, datetime.datetime]:
+        value = self.storage.get(name)
+        if isinstance(value, str):
+            decoded_value = self.decode_loaded_value(value)
+            return decoded_value
+        return value
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py]: in which cases do we store np.ndarray?

File: keepvariable/keepvariable_core.py [hunk 3]
```diff
@@ -368,10 +378,8 @@ def query(
         :rtype: list[tuple]
         """
         found_jobs: dict[str, dict] = {
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:416]: if `tag_params` or `text_params` is not None then you do this twice, what if both `tag_params` and `text_params` are specified?

File: keepvariable/keepvariable_core.py [hunk 4]
```diff
@@ -485,20 +489,39 @@ def _extract_object_from_path(self, name: str,
...
+        This command does not block the server, as it is based on a cursor-style iterator.
+
+        :param match_string: string pattern to match keys against, e.g. 'jobs:*'
+        :type match_string: str
+        :return: list of found key names
+        :rtype: list[str]
+        """
+        # Convert glob-style pattern to regex
+        match_pattern = match_string.replace("*", ".*").replace("?", ".")
+        results = [key for key in self.storage.keys() if re.search(match_pattern, key)]
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:540]: go for re.search is not None

File: keepvariable/keepvariable_core.py [hunk 5]
```diff
@@ -522,12 +545,14 @@ def set(self, key: str, value: str, additional_params: Optional[dict] = None):
...
-
-        if value is None:
-            return value
-
-        decoded_value = self.decode_loaded_value(value)
+        try:
+            value = self.redis.get(key)
+            if value is None:
+                return value
+            decoded_value = self.decode_loaded_value(value)
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:590]: could `value is None` be handled in `self.decode_loaded_value`?

File: keepvariable/keepvariable_core.py [hunk 6]
```diff
@@ -587,17 +612,42 @@ def query(
                 value_str = "|".join(values)
                 final_query += tag_query_template.format(field=field, value=value_str)
 
+        # If no query was specified, return all records from the index
+        if not final_query:
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:654]: i would compare it to empty string

File: keepvariable/keepvariable_core.py [hunk 7]
```diff
@@ -587,17 +612,42 @@ def query(
...
+
         # Query example: "@type:PIPEL @status:{QUEUED|COMPLETED}"
         # Explanation: find all jobs with type field containing 'PIPEL' and status being either 'QUEUED' or 'COMPLETED'
         query_object = Query(final_query)
         if field_to_sort_by:
             query_object.sort_by(field_to_sort_by, asc=asc)
 
         job_docs: list = self.redis.ft(index_name).search(query_object).docs
-        return {job_doc.id: job_doc.json for job_doc in job_docs}
+        return {job_doc.id: self.decode_loaded_value(job_doc.json) for job_doc in job_docs}
```
Comment (Omastto1) [inline keepvariable/keepvariable_core.py:664]: introduce `recoded_job_docs` variable for better readability

File: keepvariable/keepvariable_core.py (hunk 1 above)
Comment (DovaX) [reply keepvariable/keepvariable_core.py:227]: Both are terrible

Comment (DovaX) [reply keepvariable/keepvariable_core.py:227]: but this new one is much better on other places

File: keepvariable/keepvariable_core.py (hunk 2 above)
Comment (DovaX) [reply keepvariable/keepvariable_core.py]: Forloop doesnt, but that doesnt mean we shouldnt be able to

File: keepvariable/keepvariable_core.py (hunk 3 above)
Comment (DovaX) [reply keepvariable/keepvariable_core.py:416]: @ZaiZu2 please explain + maybe add assert if at least one of the two needs to always be None (but maybe even better would be to unify it)

File: keepvariable/keepvariable_core.py (hunk 5 above)
Comment (DovaX) [reply keepvariable/keepvariable_core.py:590]: No @Omastto1 that doesnt make sense, decode works with "not None" values which are usually dicts. None can be result here even for not-existing key - which doesnt need to be decoded at all, @ZaiZu2 please add comment to that line of code

File: keepvariable/keepvariable_core.py (hunk 6 above)
Comment (DovaX) [reply keepvariable/keepvariable_core.py:654]: Yes - makes it clear that it is always string

Comment (DovaX) [reply keepvariable/keepvariable_core.py:654]: this looks like bool

File: keepvariable/keepvariable_core.py [hunk 8]
```diff
@@ -587,17 +612,42 @@ def query(
...
+        :param match_string: string pattern to match keys against, e.g. 'jobs:*'
+        :type match_string: str
+        :param count: how many rows, defaults to 50
+        :type count: int, optional
+        :param type_: filter on specified Redis key type, defaults to None
+        :type type_: Optional[str], optional
+        :return: list of found keys
+        :rtype: list[str]
+        """
+        return list(self.redis.scan_iter(match_string, count, type_))
```
Comment (DovaX) [inline keepvariable/keepvariable_core.py:688]: Add comment here what scan_iter returns so everyone doesnt need to look it up

**Review by DovaX — APPROVED** [review]

File: keepvariable/keepvariable_core.py (hunk 3 above)
Comment (ZaiZu2) [reply keepvariable/keepvariable_core.py:416]: `query()` method is an in-memory equivalent of Redis `search()` method. In short, Redis allows you to index TEXT/TAG/... fields inside of your saved JSON documents, and use these indexes later on to perform quick searches on these fields. This method mimics this behavior, except it does not need an index to perform the search. To stay compatible with how RediSearch works, it still uses the same concepts, i.e. searching on TEXT/TAG fields. 

We can search through saved documents with any combination of TAG/TEXT params - both can be `None`, then we find all the documents under a given namespace, or both can specify some values to filter on - TEXT params will match any part of the string in the field with the string provided, while TAG params will filter on exactly matching values for the specified fields.

File: keepvariable/keepvariable_core.py (hunk 8 above)
Comment (ZaiZu2) [reply keepvariable/keepvariable_core.py:688]: I think docstring + type hinting already provides all necessary info?

**Pull request merged**